symengine/add.cpp

```diff
@@ -337,26 +337,42 @@ RCP<const Basic> Add::subs(const map_basic_basic &subs_dict) const
         return it->second;
 
     SymEngine::umap_basic_num d;
-    RCP<const Number> coef=coef_, coef2;
+    RCP<const Number> coef, coef2;
     RCP<const Basic> t;
-    for (const auto &p: dict_) {
-        RCP<const Basic> term = p.first->subs(subs_dict);
-        if (term == p.first) {
-            Add::dict_add_term(d, p.second, p.first);
-        } else if (is_a<Integer>(*term) and
-                rcp_static_cast<const Integer>(term)->is_zero()) {
-            continue;
-        } else if (is_a_Number(*term)) {
-            iaddnum(outArg(coef),
-                    mulnum(p.second, rcp_static_cast<const Number>(term)));
-        } else if (is_a<Add>(*term) and p.second->is_one()) {
+
+    auto add_cterm = [&](const RCP<const Number>& c, const RCP<const Basic>& term) {
```

Comment:
Can you move this code to a new static method like 
```
static void Add::dict_add_term_new(const Ptr<RCP<const Number>> &coef, umap_basic_num &d, 
        const RCP<const Number> &coef, const RCP<const Basic> &t);
```
This method is duplicated also in `Add::diff` and we can use it there as well.